src/bucket/BucketManager.cpp

```diff
@@ -1428,17 +1433,64 @@ loadEntriesFromBucket(std::shared_ptr<LiveBucket> b, std::string const& name,
               b->getSize(), name, ms, formatSize(bytesPerSec));
 }
 
-std::map<LedgerKey, LedgerEntry>
-BucketManager::loadCompleteLedgerState(HistoryArchiveState const& has)
+// Loads a single bucket worth of entries into `map`, deleting tombstone entries
+// and inserting archived entries. Should be called in a loop over a BL, from
+// old to new.
+static void
+loadEntriesFromHotArchiveBucket(std::shared_ptr<HotArchiveBucket> b,
+                                std::string const& name,
+                                std::map<LedgerKey, LedgerEntry>& map)
+{
+    ZoneScoped;
+
+    using namespace std::chrono;
+    medida::Timer timer;
+    HotArchiveBucketInputIterator in(b);
+    timer.Time([&]() {
+        while (in)
+        {
+            HotArchiveBucketEntry const& e = *in;
+            if (e.type() == HOT_ARCHIVE_ARCHIVED)
+            {
+                map[LedgerEntryKey(e.archivedEntry())] = e.archivedEntry();
+            }
+            else
+            {
+                if (e.type() != HOT_ARCHIVE_LIVE)
+                {
+                    std::string err =
+                        "Malformed hot archive bucket: unexpected "
+                        "non-HOT_ARCHIVE_LIVE entry.";
+                    CLOG_ERROR(Bucket, "{}", err);
+                    throw std::runtime_error(err);
+                }
+                map.erase(e.key());
+            }
+            ++in;
+        }
+    });
+    nanoseconds ns =
```

Comment:
Nit: we can further refactor timing block with:
```
  template<typename Func>
  static void
  timeBucketOperation(std::string const& name, size_t bucketSize, Func&& operation)
  {
      using namespace std::chrono;
      medida::Timer timer;
      timer.Time(std::forward<Func>(operation));

      nanoseconds ns = timer.duration_unit() * static_cast<nanoseconds::rep>(timer.max());
      milliseconds ms = duration_cast<milliseconds>(ns);
      size_t bytesPerSec = (bucketSize * 1000 / (1 + ms.count()));
      CLOG_INFO(Bucket, "Read {}-byte bucket file '{}' in {} ({}/s)",
                bucketSize, name, ms, formatSize(bytesPerSec));
  }
```

Reply:
tbh not sure saving a few lines of code is worth the extra complexity of new function + templates + forwarding